src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java

```diff
@@ -0,0 +1,274 @@
+/*
+ * RCTab
+ * Copyright (c) 2017-2023 Bright Spots Developers.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Purpose: A set of tools to verify signatures of Hart CVRs.
+ * Design: Reproduces the signature verification algorithm used by Hart, which
+ * is the .NET Framework 4.8.1 implementation of a detached XML Digital Signature.
+ * Conditions: When verifying the signature of a Hart file.
+ * Version history: see https://github.com/BrightSpots/rcv.
+ */
+
+package network.brightspots.rcv;
+
+import static network.brightspots.rcv.SecurityXmlParsers.HartSignature;
+import static network.brightspots.rcv.SecurityXmlParsers.RsaKeyValue;
+import static network.brightspots.rcv.SecurityXmlParsers.SignedInfo;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.dataformat.xml.XmlMapper;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.RSAPublicKeySpec;
+import java.util.Base64;
+import javax.xml.crypto.OctetStreamData;
+import javax.xml.crypto.dsig.CanonicalizationMethod;
+import javax.xml.crypto.dsig.TransformException;
+import javax.xml.crypto.dsig.XMLSignatureFactory;
+import javax.xml.crypto.dsig.spec.C14NMethodParameterSpec;
+
+class SecuritySignatureValidation {
+  /**
+   * Throws an exception if the signature is invalid.
+   * Throws CouldNotVerifySignature if it is unable to run the validation, for reasons including:
+   * 1. The public key in the signed file does not match the expected key
+   * 2. The canonicalization or hashing algorithm is not supported
+   * 3. The data file has been edited and its hash does not match what's found in signatureKeyFile
+   * Throws VerificationFailedException if the signature verification successfully ran, but
+   * the corresponding signature was invalid.
+   */
+  public static void ensureSignatureIsValid(
+          RsaKeyValue publicKey,
+          File signatureKeyFile,
+          File dataFile) throws CouldNotVerifySignatureException, VerificationFailedException {
```

Comment:
The distinction between this two exception types is a bit unclear given the names you chose. Maybe something like `VerificationUnableToRunException` and `VerificationSignatureDidNotMatchException`?

Reply:
Agreed.

What I'm trying to get at is that it's possible for the signature to be correct, but we didn't even check the signature because there were other errors first. For example, the signature could be correct, but the key was different; or the signature was correct, but the file hash was different.

For that reason, I'll use `VerificationDidNotRun` and `VerificationSignatureDidNotMatch`